extensions/s3/src/main/java/io/deephaven/extensions/s3/S3RequestCache.java

```diff
@@ -57,36 +57,37 @@ int getFragmentSize() {
      * @return the request
      */
     @NotNull
-    S3ChannelContext.Request.AcquiredRequest getOrCreateRequest(@NotNull final S3Uri uri, final long fragmentIndex,
+    Request.AcquiredRequest getOrCreateRequest(@NotNull final S3Uri uri, final long fragmentIndex,
             @NotNull final S3ChannelContext context) {
         final Request.ID key = new Request.ID(uri, fragmentIndex);
         Request.AcquiredRequest newAcquiredRequest = null;
-        Request existingRequest = requests.get(key);
-        for (int retryCount = 0; retryCount < Integer.MAX_VALUE; retryCount++) {
+        while (true) {
+            final Request existingRequest = requests.get(key);
             if (existingRequest != null) {
                 final Request.AcquiredRequest acquired = existingRequest.tryAcquire();
                 if (acquired != null) {
                     return acquired;
-                } else {
-                    remove(existingRequest);
                 }
             }
             if (newAcquiredRequest == null) {
                 newAcquiredRequest = Request.createAndAcquire(fragmentIndex, context);
             }
-            existingRequest = requests.putIfAbsent(key, newAcquiredRequest.request);
+            final boolean added;
             if (existingRequest == null) {
+                // Ideally, we could have used ".replace" in this case as well, but KeyedObjectHashMap.replace currently
+                // has a bug when the key is not present in the map.
+                added = requests.putIfAbsent(key, newAcquiredRequest.request) == null;
+            } else {
+                added = requests.replace(key, existingRequest, newAcquiredRequest.request);
```

Comment:
This is good if we can convince ourselves that there is no bug in `replace`.

Reply:
Otherwise we need something like what you had:
```
        Request existingRequest = requests.get(key);
        while (true) {
            if (existingRequest != null) {
                final Request.AcquiredRequest acquired = existingRequest.tryAcquire();
                if (acquired != null) {
                    return acquired;
                } else {
                    remove(existingRequest);
                }
            }
            if (newAcquiredRequest == null) {
                newAcquiredRequest = Request.createAndAcquire(fragmentIndex, context);
            }
            if ((existingRequest = requests.putIfAbsent(key, newAcquiredRequest.request)) == null) {
```